### ISSUE TITLE
Formal AttnLRP Verification / Outline for using LXT properly?

### ISSUE THREAD

**Issue by andrewboldi** [rachtibat/LRP-eXplains-Transformers#32]
Hi, thank you for making this wonderful repo!

I am interested in implement AttnLRP for a custom architecture and I just wanted to confirm that I am understanding the steps correctly. From my understanding, all I need to do is:

1. Manually edit the gradient computation for non-linear activations, element-wise matrix multiplications, matrix multiplications, and normalization operations according to the examples at https://lxt.readthedocs.io/en/latest/extending.html with identity_rule_implicit, divide_gradient, divide_gradient, and stop_gradient, respectively.
2. Ensure gradient computation is enabled for the input embeddings.
3. Call .backward() on the target logit.
4. Multiply the input embedding gradients by the input embedding element-wise.

Would this be a correct method? And is there a way I can (formally) verify that the relevance I obtained is accurately computed (and I didn't accidentally forget about a matrix multiplication somewhere)?

Also, since I am modifying the architecture itself (which might also be later used for training), do you have any recommendations for LRP computation without breaking normal gradient computation (during training) other than keeping two separate models or just including a parameter for training/LRP calculation?

Thank you very much in advance!

Kind regards,
Andrew

**Comment by ja-nina** [comment]
Also highly interested in this :)

**Comment by rachtibat** [comment]
Hey @andrewboldi and @ja-nina,

exactly. So, please
1. apply identity rule with identity_rule_implicit on non-linearities in MLP blocks
2. if you have eager attention, divide the gradient by 2 before you apply any torch.matmul operation with divide_gradient 
4. if you have *standard flash attention without softcapping*, you can divide the gradients at value, key, query like here: https://github.com/rachtibat/LRP-eXplains-Transformers/blob/99c0b5566cbd9e5acd0250e6e98cda5f94fa244d/lxt/efficient/patches.py#L196
5. flash attention with softcapping is not supported. Please use eager attention and then apply the identity rule on the tanh. This attention is only found in gemma models. I will hopefully soon find time to include gemma models in LXT.
6. stop the gradient flow in the normalization operation in LayerNorm like here https://github.com/rachtibat/LRP-eXplains-Transformers/blob/99c0b5566cbd9e5acd0250e6e98cda5f94fa244d/lxt/efficient/patches.py#L136 using stop_gradient.
7. in element-wise multiplications, use the divide_gradient by 2 operator like in https://github.com/rachtibat/LRP-eXplains-Transformers/blob/99c0b5566cbd9e5acd0250e6e98cda5f94fa244d/lxt/efficient/patches.py#L156
8. Enjoy!

of course activate also gradients and backward only the target logit.

There is no way to test if you implementation is correct inside LXT right now. Usually, you can simply look at a heatmap and see if it makes sense. If you want to be 100% sure, @ja-nina made a fantastic suggestion. You can test if the conservation property of LRP holds. I.e. if the sum of relevance in the input embeddings = target logit value, that you explained.
This only works if you disable all bias terms in all linear layers ~~and also replace the softmax with nn.Identity just for the sake of testing conservation.~~ _(EDIT: Another better way to do it would be to `.detach()` the softmax from the computational graph. So you would write `softmax(x).detach()`. This will keep numerical stability and force all the gradients to flow through the values instead and keep conservation. Also don't divide the gradients by 2 after the softmax matmul anymore, because now the relevance flows only through one path as if the softmax is a constant weight.)_
@ja-nina this might be the reason why you did not get perfect conservation because the softmax is eating some relevance!

If you want to train, you cant use this same script unfortunately. You have two choices: 

1. Either modify your source code with some sort of flag, where you can globally toggle on and off the modifications. For that I would not use the monkey patch logic, but modify the source code directly.
2. You need to have two separate huggingface transformers definition classes, where you load in the same weights. So, you have the weights only loaded once but two different model definitions. Since, we monkey patch the transformers library, it might make sense, if you copy the class definition file, rename it and then import it. Also make sure the monkey patch doesnt patch PyTorch classes. Because here I patch all torch.nn.Dropout classes e.g. https://github.com/rachtibat/LRP-eXplains-Transformers/blob/main/lxt/efficient/models/llama.py, which means that your second model will also use this modified dropout! So, you also need to define a custom Dropout and delete this monkey patching. I hope this was a bit clear. 

So bottom line is: Make sure that the monkey patch does not patch your model definition that you use for training!

Best,
Reduan

**Comment by andrewboldi** [comment]
Hi @rachtibat, thank you so much for your detailed reply! It was very helpful and after implementing all the steps, I obtained a visualization for my model that looked great! 

Thank you also for your suggestion about using the same class definitions for training a model. I ended up manually including a flag/argument throughout the architecture for matrix multiplications and used `monkey_patch` with a custom `patch_map` (for `nn.LayerNorm` and all activation functions). 

I still have a coupling lingering questions:

1. As a quick sanity check, would it be true that positive/negative values and higher/lower values from the relevance we obtain from normalizing (input * input.grad) represent inputs that correspond more strongly/contribute contradictorily to the target class, respectively?
2. In my architecture, I have a batch matrix multiplication (`torch.bmm`). My reasoning is to still use `divide_gradient(res, 2)` since each matrix across the 0-th dimension is independent of the other. In other words, it should be equivalent to iterating across each 3-d tensor, performing matrix multiplication, and dividing the gradient by 2 on each product. Does this seem reasonable?
3. Your and @ja-nina's suggestions about checking conservation by zeroing the biases like in zennit is a great idea! I tried following what you suggested @rachtibat (setting the bias to None for all nn.Linear layers and excluding the softmax function) yet for some reason the sum of the relevances across the input embeddings is on the order of $10^{9}$ (when my logit is like $2.04$. For reference, I am still monkey patching all the Layer Norms with `layer_norm_forward` and all the activation functions with `identity_rule_implicit` (it is also quite off even when I disable monkey patching). I also tried disabling the LayerNorm biases, tweaking the epsilon in layer_norm_forward, and confirmed all biases are disabled globally with `torchinfo` to no avail. I'm thinking that there is some sort of numerical instability potentially arising from having no bias term in the Linear layers which leads to runaway relevances. Also, for reference, I noticed that if I do not disable bias terms/softmax, the sum of the relevance in the input embedding is about 0.3 (1 order of magnitude lower than the logit) which suggests around 90% of the relevance might be lost to non-linearities. I think this latter point makes sense. In your experience, is this reasonable? Also, do you have any ideas why conservation is not being preserved and how I could maybe fix it?

Thank you very much for all your support and excellent paper!

**Comment by rachtibat** [comment]
Hey @andrewboldi,

I'm glad that you found my tipps helpful (:

Regarding your questions:

1. Yes! So, the more positive a value, the stronger it contributes positively to the prediction logit. The more negative, the more suppressive.
2. Yes, makes sense! I would also divide the gradient by two. But only if you have a bi-linear matrix multiplication. This means, that both inputs to torch.bmm are also input variables you want to explain. E.g. doing torch.matmul(x, W) where W are model weights, here we would NOT apply the divide 2 rule, because the weights W do not receive relevance, they are a passive element of the network. But in attention layers, we multiply the queries and keys. That's why we need to compute relevance for both inputs and also divide by 2.
3. You're right again. I might have done a mistake suggesting to you to replace the softmax with nn.Identity... I actually have a section in our paper https://arxiv.org/pdf/2402.05602 page 22, where I show applying the identity rule on softmax creates numerical instabilities.
Another better way to do it would be to `.detach()` the softmax from the computational graph. So you would write `softmax(x).detach()`. This will keep numerical stability and force all the gradients to flow through the values instead and keep conservation. Also don't divide the gradients by 2 after the softmax matmul anymore, because now the relevance flows only through one path as if the softmax is a constant weight.

So in your case, you have 0.3 relevance left, which makes sense! This is normal due to softmax and linear layer biases.

Best

**Comment by andrewboldi** [comment]
Hi @rachtibat, thank you very much again for your detailed reply! Thank you for the clarifications on the relevance magnitudes and the explanation of the difference between bilinear and linear matrix multiplication! Thank you also for the clarification about the `softmax(x).detach()`! This completely solved the numerical instability problem that I was having!

After implementing all the changes, I realized that my architecture still had an _implicit_ linear layer because torch.nn.LayerNorm by default includes a bias term. After manually disabling the bias terms in all the LayerNorms, Linear layers, and applying `softmax(x).detach()`, the sum of the relevance at the input layer is exactly equal to my output logit!

Thank you again for all your help!
